### PR TITLE
Do not crash if remote file does not exist

### LINKED ISSUE #208: pylibssh takes surprisingly longer time than paramiko to receive the same data
##### SUMMARY
- Related to https://github.com/ansible-collections/cisco.nxos/issues/145
- The above issue is caused because pylibssh takes ~300 times longer than paramiko to read the output of `show vlan | json` from the same target device. This resulted in the `command_timeout` value to work for paramiko but not libssh.
- This was realised by adding `time.perf_counter()` in relevant places within network_cli code as defined in this [patch](https://gist.github.com/NilashishC/47d58086f31ceda5c9e52c5f39c212a0).
- Based on multiple passes, paramiko takes ~3 secs while pylibssh takes ~1000 secs to receive the same data.
- Data from one such pass is added here - https://gist.github.com/NilashishC/62cf8e945ac9a49e743fb2b553b8716d.
- On bumping the buffer size in `read_non_blocking()` to `8096` bytes and removing the upper cap of `1024` bytes, this delay was reduced to ~280 secs.

##### ISSUE TYPE
- Bug Report

##### PYLISSH and LIBSSH VERSION

```console
$ pip show ansible-pylibssh
Name: ansible-pylibssh
Version: 0.2.0
Summary: Python bindings for libssh client specific to Ansible use case
Home-page: https://github.com/ansible/pylibssh
Author: Ansible, Inc.
Author-email: info+github/ansible/pylibssh@ansible.com
License: LGPLv2+
Location: /Users/nchakrab/venvs/dev/lib/python3.6/site-packages
Requires:
Required-by:
```
```pycon
>>> import pylibsshext
>>> pylibsshext.__libssh_version__
'0.9.5'
```

##### OS / ENVIRONMENT
ProductName:    Mac OS X
ProductVersion: 10.15.7

##### STEPS TO REPRODUCE
- Configure 500+ vlans on a Cisco Nexus appliance.
- Run the following playbook with first paramiko and then libssh.
```yaml
---
- hosts: nxos
  gather_facts: no
  tasks:
    - cisco.nxos.nxos_vlans:
        config:
          - vlan_id: 10
            enabled: True
        state: merged
```
        
##### EXPECTED RESULTS
The playbook run doesn't take an exorbitantly high time to complete.

##### ACTUAL RESULTS
When `command_timeout < 1100` it results in a timeout. With a higher `command_timeout` set, the playbook run completes as expected.

### LINKED ISSUE #325: Unable to transfer files using SCP to a Cisco IOS device
##### SUMMARY
When attempting to transfer a file using SCP to an IOS device, you get a "No such file or directory" error. If you switch back to Paramiko it works fine.

##### ISSUE TYPE
- Bug Report

##### PYLISSH and LIBSSH VERSION

```
libssh_0.9.3
ansible_pylibssh-0.4.0
```

##### OS / ENVIRONMENT

This is running from a custom EE based off of `quay.io/ansible/awx-ee:0.6.0`. The only additions are some extra packages for netbox, scp, and pylibssh.

The target is a Cisco 3750xp running IOS 15.2(4)E8. I have run this with ssh1 and ssh2 enabled on the switch, and both ended with the same result.

##### STEPS TO REPRODUCE

Running this task

```
- name: transfering upgrade image to switch...
  block:
    - ansible.netcommon.net_put:
        src: "../files/c3750e-universalk9-mz.152-4.E10.bin"
        dest: "flash:c3750e-universalk9-mz.152-4.E10.bin"
  when: ansible_network_os == 'cisco.ios.ios'
```

##### EXPECTED RESULTS

This is actually the output when you add the var `ansible_network_cli_ssh_type: paramiko`, nothing else was changed between this run and the failed one using libssh.

SSH Logs from the switch: https://gist.github.com/vrelk/da99ec3107e18d7239f0286ee582e2c9 (start/end trimmed)

```
TASK [upgrade : ansible.netcommon.net_put] *************************************
task path: /runner/project/roles/upgrade/tasks/xfer.yml:3
redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
redirecting (type: become) ansible.builtin.enable to ansible.netcommon.enable
<10.255.79.10> attempting to start connection
<10.255.79.10> using connection plugin ansible.netcommon.network_cli
Found ansible-connection at path /usr/local/bin/ansible-connection
<10.255.79.10> found existing local domain socket, using it!
<10.255.79.10> updating play_context for connection
<10.255.79.10> 
<10.255.79.10> local domain socket path is /home/runner/.ansible/pc/11ab5503fd
Getting network OS from inventory
changed: [3750xp-stack] => {
    "changed": true,
    "destination": "flash:c3750e-universalk9-mz.152-4.E10.bin"
}
```

##### ACTUAL RESULTS

This is the debug output on the switch that corresponds to this task (with ssh and scp debugging enabled): https://gist.github.com/vrelk/23fa0dfd09b8f27f79118684bac75d08

```
TASK [upgrade : ansible.netcommon.net_put] *************************************
task path: /runner/project/roles/upgrade/tasks/xfer.yml:3
redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
redirecting (type: become) ansible.builtin.enable to ansible.netcommon.enable
<10.255.79.10> attempting to start connection
<10.255.79.10> using connection plugin ansible.netcommon.network_cli
Found ansible-connection at path /usr/local/bin/ansible-connection
<10.255.79.10> found existing local domain socket, using it!
<10.255.79.10> updating play_context for connection
<10.255.79.10> 
<10.255.79.10> local domain socket path is /home/runner/.ansible/pc/11ab5503fd
Getting network OS from inventory
fatal: [3750xp-stack]: FAILED! => {
    "changed": false,
    "destination": "flash:c3750e-universalk9-mz.152-4.E10.bin",
    "msg": "Exception received: Error transferring file to flash:c3750e-universalk9-mz.152-4.E10.bin: Initializing SCP session of remote file [flash:c3750e-universalk9-mz.152-4.E10.bin] for write failed with error [b'Socket error: No such file or directory']"
}
```

### PR DESCRIPTION
**Pull request by Jakuje**
##### SUMMARY
The uninitialized pointer caused pylibssh crashing when the `scp.get()` failed in the first invocation of  `libssh.ssh_scp_pull_request()` (generally when the remote file did not exist, or it was not possible to read it for some reason).

This is the full log from libssh when pytest is executed directly with the `-s` argument:
```console
[2024/06/21 16:04:37.620775, 2] ssh_scp_pull_request:  Received SCP request: 'scp: /tmp/pytest-of-jjelen/pytest-22/popen-gw1/test_get_missing_src0/non-existing.txt: No such file or directory'
[2024/06/21 16:04:37.620779, 1] ssh_scp_pull_request:  SCP: Warning: scp: /tmp/pytest-of-jjelen/pytest-22/popen-gw1/test_get_missing_src0/non-existing.txt: No such file or directory
free(): invalid pointer
Fatal Python error: Aborted
```
(running the test from tox showed just the last line with pytest crashing)

Fixes #208
Fixes #325

##### ISSUE TYPE
- Bugfix Pull Request

### CODE REVIEW

File: docs/changelog-fragments/620.bugfix.rst
Comment (webknjaz) [inline docs/changelog-fragments/620.bugfix.rst]: Could you use the past tense or a non-imperative present simple with the reference to what's changed since the previous release?

Comment (Jakuje) [reply docs/changelog-fragments/620.bugfix.rst]: Something like this works?
```
Fixed crash when downloading non-existent remote file using SCP -- by :user:`Jakuje`
```
(avoiding push to let the CI run)

File: docs/changelog-fragments/620.bugfix.rst
Comment (webknjaz) [inline docs/changelog-fragments/620.bugfix.rst]: This would work. I think that it could also be something like
```suggestion
Downloading non-existent remote files via SCP no longer crashes the program -- by :user:`Jakuje`
```

File: tests/conftest.py
Comment (webknjaz) [inline tests/conftest.py]: Let's move this to a separate PR, as it's not related to the segfault. It'll need to come in through a fixture interface and avoid relying on env vars. That fixture will need to read pytest's own verbosity level instead of having an additional toggle for the same thing, similar to this example: https://docs.pytest.org/en/stable/reference/reference.html#pytest.fixtures.pytestconfig.

**Comment by webknjaz** [comment]
Overall, I like the fix and the test. Let's move the unrelated stuff out and merge it. I hope this will also address the flakiness of some tests...

File: docs/changelog-fragments/620.bugfix.rst
Comment (Jakuje) [reply docs/changelog-fragments/620.bugfix.rst]: Updated as suggested.

File: tests/conftest.py
Comment (Jakuje) [reply tests/conftest.py]: Removed. Thanks for suggestion. I will look into this in separate PR.

**Comment by webknjaz** [comment]
@Jakuje could you also check if this is going to help with #222 / #208?

File: docs/changelog-fragments/620.bugfix.rst
Comment (webknjaz) [inline docs/changelog-fragments/620.bugfix.rst]: I suppose I forgot to clarify what counts as the end of a sentence :)
```suggestion
Downloading non-existent remote files via SCP no longer crashes the program -- by :user:`Jakuje`.
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py:68]: oh! that's where #621 got the misleading fixture from :)

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
def path_to_non_existent_src_file(tmp_path):
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
def test_get_missing_src(dst_path, path_to_non_existent_src_file, ssh_scp):
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
        ssh_scp.get(str(path_to_non_existent_src_file), str(dst_path))
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]: Since the destination path shouldn't influence the problem, I'd stop relying on that fixture.
```suggestion
        ssh_scp.get(str(path_to_non_existent_src_file), os.devnull)
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
def test_get_missing_src(path_to_non_existent_src_file, ssh_scp):
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py:7]:
```suggestion
import os

```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
import os
import uuid
```

File: tests/unit/scp_test.py
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
def test_copy_from_non_existent_remote_path(path_to_non_existent_src_file, ssh_scp):
```

File: docs/changelog-fragments/620.bugfix.rst
Comment (webknjaz) [inline docs/changelog-fragments/620.bugfix.rst:1]: Could you also add a symlink named `325.bugfix.rst` pointing to this file? This will link the same entry to both issues/PRs.

**Review by webknjaz — APPROVED** [review]
@Jakuje I just realized that we need to link the change note to #325 as well, and I wanted to give you a chance to clean up the branch after my changes. So I'm approving the PR but will wait for the final touches from you before hitting merge.

**Comment by Jakuje** [comment]
> @Jakuje could you also check if this is going to help with #222 / #208?

Likely the #208 but most likely not the #222 as the SFTP goes through different code path as far as I know. I will squash the fixups now.

File: docs/changelog-fragments/620.bugfix.rst
Comment (Jakuje) [reply docs/changelog-fragments/620.bugfix.rst:1]: Done. Also for the 208.

File: tests/unit/scp_test.py
Comment (Jakuje) [reply tests/unit/scp_test.py:68]: sorry. Likely messed some parts together when I was testing both of them.

**Comment by Jakuje** [comment]
Should be "cleaned up". Let me know if there would be some other issues.

**Review by KB-perByte — APPROVED** [review]
Thanks for fixing this @Jakuje

File: tests/unit/scp_test.py
Comment (webknjaz) [reply tests/unit/scp_test.py:68]: No problem! That was mostly a note to myself — I tend to record my discoveries, as it helps to build a fuller picture of what patches do during reviews…

File: docs/changelog-fragments/620.bugfix.rst
Comment (webknjaz) [reply docs/changelog-fragments/620.bugfix.rst:1]: Wonderful! Thanks!

**Comment by webknjaz** [comment]
@Jakuje thank you! I enabled auto-merge, it'll get into `devel` once the CI completes.

File: docs/changelog-fragments/620.bugfix.rst
Comment (webknjaz) [reply docs/changelog-fragments/620.bugfix.rst:1]: FTR, here's the changelog preview: https://ansible-pylibssh--620.org.readthedocs.build/en/620/changelog/#bug-fixes.

**Comment by webknjaz** [comment]
@Jakuje @KB-perByte this has been released as a part of v1.2.1: https://github.com/ansible/pylibssh/releases/tag/v1.2.1 / https://github.com/ansible/pylibssh/discussions/613 / https://pypi.org/project/ansible-pylibssh/1.2.1/.

Enjoy! :tada:

**Comment by Jakuje** [comment]
Thank you for the patience in getting this to the standard :)

**Comment by webknjaz** [comment]
No problem ;) I actually enjoy helping contributors who maintain such a quality in their PRs.

**Pull request merged**
